File: IPC/SharedMemory/C/shm_example.c
#include <sys/shm.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/wait.h>

int main() {
    // Generate unique key for shared memory
    key_t key = ftok("/tmp", 'C');
    // Allocate shared memory segment using shmget() - creates segment of 1024 bytes
    int shmid = shmget(key, 1024, 0666 | IPC_CREAT);
    // Attach shared memory to process address space using shmat()
    // NULL: let system choose address, 0: read/write permissions
    char *str = (char*) shmat(shmid, NULL, 0);

    pid_t pid = fork();
    if (pid == 0) { // Child
        printf("Child reads: %s\n", str);
        // Detach shared memory using shmdt() - removes from address space
        shmdt(str);
    } else { // Parent
        strcpy(str, "Hello from parent");
        sleep(1); // Wait for child to read
        // Detach shared memory
        shmdt(str);
        // Remove shared memory segment using shmctl() with IPC_RMID
        shmctl(shmid, IPC_RMID, NULL);
        wait(NULL);
    }

    return 0;
}

File: IPC/SharedMemory/Python/shm_example.py
import mmap
import os
import time

def child():
    # Open the shared memory file
    fd = os.open('/tmp/shared_memory', os.O_RDONLY)
    # Memory map the file
    mm = mmap.mmap(fd, 0, mmap.MAP_SHARED, mmap.PROT_READ)
    print("Child reads:", mm.read().decode().strip())
    mm.close()
    os.close(fd)

if __name__ == '__main__':
    # Create a shared memory file
    with open('/tmp/shared_memory', 'w+b') as f:
        f.write(b'Hello from parent')

    # Fork the process
    pid = os.fork()
    if pid == 0:
        # Child process
        child()
    else:
        # Parent process
        time.sleep(1)  # Wait for child to read
        os.waitpid(pid, 0)  # Wait for child to finish